src/Cqm/QrdaControllers/QrdaReportController.php

```diff
@@ -69,41 +66,71 @@ public function getCategoryIReport($pid, $measures, $type = 'xml')
             }
             chmod($zip_directory, 0777);
         }
         $pids = is_array($pids) ? $pids : [$pids];
-        foreach ($measures as $measure) {
-            $measure_directory = $zip_directory . "/" . $measure . "_" . time();
-            if (!is_dir($measure_directory)) {
-                if (!mkdir($measure_directory, true) && !is_dir($measure_directory)) {
-                    throw new \RuntimeException(sprintf('Directory "%s" was not created', $measure_directory));
+        if (!$bypid) {
+            foreach ($measures as $measure) {
+                if (is_array($measure)) {
+                    $dir_measure = $measure['measure_id'];
+                } else {
+                    $dir_measure = $measure;
+                }
+                $measure_directory = $zip_directory . "/" . $dir_measure . "_" . time();
+                if (!is_dir($measure_directory)) {
+                    if (!mkdir($measure_directory, true) && !is_dir($measure_directory)) {
+                        throw new \RuntimeException(sprintf('Directory "%s" was not created', $measure_directory));
+                    }
+                    chmod($measure_directory, 0777);
+                }
+                foreach ($pids as $pid) {
+                    $meta = sqlQuery("Select `fname`, `lname` From `patient_data` Where `pid` = ?", [$pid]);
+                    $file = $measure_directory . "/{$meta['fname']}_{$meta['lname']}." . $type;
+                    $content = $this->getCategoryIReport($pid, $measure, $type);
+                    $f_handle = fopen($file, "w");
+                    fwrite($f_handle, $content);
+                    unset($content);
+                    unset($file);
+                    fclose($f_handle);
+                    if ($type === 'xml') {
+                        copy(__DIR__ . '/../../../interface/modules/zend_modules/public/xsl/qrda.xsl', $measure_directory . "/qrda.xsl");
+                    }
                 }
-                chmod($measure_directory, 0777);
             }
+            $zip_name = "QRDA1_Export_" . time() . ".zip";
+            $save_path = sys_get_temp_dir() . "/" . $zip_name;
+            $zip->setArchive($save_path);
+            $zip->compress($zip_directory);
+        } elseif ($bypid) {
             foreach ($pids as $pid) {
                 $meta = sqlQuery("Select `fname`, `lname` From `patient_data` Where `pid` = ?", [$pid]);
-                $file = $measure_directory . "/{$meta['fname']}_{$meta['lname']}." . $type;
-                $content = $this->getCategoryIReport($pid, $measure, $type);
-                $directory = $GLOBALS['OE_SITE_DIR'] . DIRECTORY_SEPARATOR . 'documents' . DIRECTORY_SEPARATOR . 'temp';
-                file_put_contents($directory . DIRECTORY_SEPARATOR . $measure . "_{$meta['lname']}_{$meta['fname']}." . 'xml', $content);
+                $file = $zip_directory . "/{$meta['fname']}_{$meta['lname']}." . $type;
```

Comment:
potential issue here where fname or lname could be set to traverse directories (in cases like this, would run those variables through the convert_safe_file_dir_name() function which is always available in the library/sanitize.inc.php script)